File: Sort Colors/sort-colors.py
# This problem is also known as the Dutch national flag problem.

from typing import List

class Solution:
    def sortColors(self, nums: List[int]) -> List[int]:

        zero = 0 # all elements before nums[zero] are 0's
        curr = 0 # pointer to a current element of array
        two = len(nums) - 1 # all elements after nums[two] are 2's

        while curr <= two:
            if nums[curr] == 0:
                nums[curr], nums[zero] = nums[zero], nums[curr]
                zero += 1
                curr += 1
            elif nums[curr] == 1:
                curr += 1
            else:
                nums[curr], nums[two] = nums[two], nums[curr]
                two -= 1

        return nums

File: Sort Colors/sort-colors.cpp
class Solution {
public:
  void sortColors(vector<int>& nums) {
    
    int zero = 0;
    int curr = 0;
    int two = nums.size() - 1;

    while (curr <= two) {
      if (nums[curr] == 0) {
	std::swap(nums[curr], nums[zero]);
	++zero;
	++curr;
      }
      else if (nums[curr] == 1)
	++curr;
      else {
	std::swap(nums[curr], nums[two]);
	--two;
      }
    }
  }
};
